javaslang/src/main/java/javaslang/collection/Iterator.java

```diff
@@ -2089,6 +2063,72 @@ public String toString() {
         }
     }
 
+    final class GroupedIterator<T> extends AbstractIterator<Seq<T>> {
+
+        private final Iterator<T> that;
+        private final int size;
+        private final int step;
+        private final int gap;
+        
+        private Seq<T> buffer = Vector();
+        private boolean filled = false;
+
+        GroupedIterator(Iterator<T> that, int size, int step) {
+            if (size < 1 || step < 1) {
+                throw new IllegalArgumentException("size (" + size + ") and step (" + step + ") must both be positive");
+            }
+            this.that = that;
+            this.size = size;
+            this.step = step;
+            this.gap = Math.max(step - size, 0);
+        }
+
+        @Override
+        public boolean hasNext() {
+            return filled || fill();
+        }
+
+        @Override
+        protected Seq<T> getNext() {
+            filled = false;
+            return buffer;
+        }
+
+        private boolean fill() {
```

Comment:
this seems very stateful, would it be possible to separate it into smaller, more focused methods?